Add game_template

filiph opened the pull request:
Adds a template / sample for games built in Flutter, with all the bells and whistles, like ads, in-app purchases, audio, main menu, settings, and so on.

// cc @domesticmouse @craiglabenz 

## Pre-launch Checklist

- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I read the [Contributors Guide].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.

domesticmouse commented on game_template/macos/Runner/Release.entitlements:
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>com.apple.security.app-sandbox</key>
```
Does the app need network access in prod?

filiph replied:
It does, for Crashlytics.

domesticmouse reviewed (approved):
This is going to need a follow up PR that fixes up lints for Dart 2.17

parlough commented on game_template/lib/src/audio/sounds.dart:
```diff
@@ -0,0 +1,67 @@
+List<String> soundTypeToFilename(SfxType type) {
```
Maybe adjust these lists to be `const`.

filiph replied:
Will do.

parlough commented on game_template/lib/src/crashlytics/crashlytics.dart:
```diff
@@ -0,0 +1,96 @@
+/// https://stackoverflow.com/questions/47654410/how-to-effectively-group-non-fatal-exceptions-in-crashlytics-fabrics.
+@visibleForTesting
+StackTrace filterStackTrace(StackTrace stackTrace) {
+  StackTrace? filtered;
+  try {
+    final lines = stackTrace.toString().split('\n');
+    final buf = StringBuffer();
+    for (final line in lines) {
+      if (line.contains('crashlytics.dart') ||
+          line.contains('_BroadcastStreamController.java') ||
+          line.contains('logger.dart')) {
+        continue;
+      }
+      buf.writeln(line);
+    }
+    filtered = StackTrace.fromString(buf.toString());
+  } catch (e) {
+    debugPrint('Problem while filtering stack trace: $e');
+  }
+
+  return filtered ?? stackTrace;
```
Can just skip the variable:

```suggestion
  try {
    final lines = stackTrace.toString().split('\n');
    final buf = StringBuffer();
    for (final line in lines) {
      if (line.contains('crashlytics.dart') ||
          line.contains('_BroadcastStreamController.java') ||
          line.contains('logger.dart')) {
        continue;
      }
      buf.writeln(line);
    }
    return StackTrace.fromString(buf.toString());
  } catch (e) {
    debugPrint('Problem while filtering stack trace: $e');
  }
  
  // If there was an error while filtering,
  // return the original, unfiltered stack track.
  return stackTrace;
```

filiph replied:
Thanks, you opened my mind here. All this time I was avoiding control statements anywhere in `try`/`catch`, but it turns out it only makes sense to avoid them in `finally`, and it's completely fine to have them in `try`.

parlough commented on game_template/lib/src/in_app_purchase/in_app_purchase.dart:
```diff
@@ -0,0 +1,189 @@
+
+  Future<void> _listenToPurchaseUpdated(
+      List<PurchaseDetails> purchaseDetailsList) async {
+    for (var purchaseDetails in purchaseDetailsList) {
```
Maybe make this `final` to be consistent with other loops in the sample:

```suggestion
    for (final purchaseDetails in purchaseDetailsList) {
```

parlough commented on game_template/README.md:
````diff
@@ -0,0 +1,538 @@
+flutter build appbundle && open build/app/outputs/bundle/release
+```
+
+While the template is meant for mobile games, you can also publish
````
As a reader this comment makes me think the sample while not intended for web will at least be buildable/runnable. I can see confusion popping up when the sample won't work. 

Perhaps we can clarify that is not the case for this specific sample or that more changes would be necessary?

filiph replied:
Ah, thanks for mentioning it. The game _is_ runnable on the web, but I had a bug with the Crashlytics code that would crash the game on startup in the browser.

It's fixed now.

parlough replied:
Even better, thanks!

parlough commented on game_template/lib/src/audio/audio_controller.dart:
```diff
@@ -0,0 +1,265 @@
+            (i) => AudioPlayer(
+                playerId: 'sfxPlayer#$i',
+                mode: PlayerMode.LOW_LATENCY)).toList(growable: false),
+        _playlist = Queue.from(List<Song>.of(songs)..shuffle()) {
```
```suggestion
        _playlist = Queue.of(List<Song>.of(songs)..shuffle()) {
```

parlough commented on game_template/lib/src/audio/songs.dart:
```diff
@@ -0,0 +1,20 @@
+const Set<Song> songs = {
+  // Filenames with whitespace break package:audioplayers on iOS
+  // (as of February 2022), so we use no whitespace.
+  Song('Mr_Smith-Azul.mp3', 'Azul'),
+  Song('Mr_Smith-Sonorus.mp3', 'Sonorus'),
+  Song('Mr_Smith-Sunday_Solitude.mp3', 'SundaySolitude'),
+};
+
+class Song {
+  final String filename;
+
+  final String author = 'Mr Smith';
+
+  final String name;
+
+  const Song(this.filename, this.name);
+
+  @override
+  String toString() => 'Song<$filename>';
+}
```
Maybe adjust `author` to `artist` with an optional initializing parameter to make the class slightly more flexible/generic.

```suggestion
const Set<Song> songs = {
  // Filenames with whitespace break package:audioplayers on iOS
  // (as of February 2022), so we use no whitespace.
  Song('Mr_Smith-Azul.mp3', 'Azul', artist: 'Mr Smith'),
  Song('Mr_Smith-Sonorus.mp3', 'Sonorus', artist: 'Mr Smith'),
  Song('Mr_Smith-Sunday_Solitude.mp3', 'SundaySolitude', artist: 'Mr Smith'),
};

class Song {
  final String filename;

  final String name;

  final String? artist;

  const Song(this.filename, this.name, {this.artist});

  @override
  String toString() => 'Song<$filename>';
}
```

parlough reviewed (approved):
This is really cool to see. Going through the sample, I learned a lot about some of the packages from your comments and usage, thanks!

I have some minor nits since this seems slightly intended to be used as a template, feel free to consider them or not :)

filiph commented on game_template/lib/src/audio/audio_controller.dart:
```diff
@@ -0,0 +1,265 @@
+      return;
+    }
+
+    if (_settings != null) {
+      _settings!.muted.removeListener(_mutedHandler);
+      _settings!.musicOn.removeListener(_musicOnHandler);
+      _settings!.soundsOn.removeListener(_soundsOnHandler);
+    }
+
+    _settings = settingsController;
+
+    _settings!.muted.addListener(_mutedHandler);
+    _settings!.musicOn.addListener(_musicOnHandler);
+    _settings!.soundsOn.addListener(_soundsOnHandler);
+
+    if (!_settings!.muted.value && _settings!.musicOn.value) {
+      _startMusic();
+    }
```
That works, thanks!

filiph commented on game_template/lib/src/audio/sounds.dart:
```diff
@@ -0,0 +1,67 @@
+  }
+}
+
+enum SfxType {
```
Yes!

filiph commented on game_template/lib/src/audio/audio_controller.dart:
```diff
@@ -0,0 +1,265 @@
+  /// goes into the background.
+  void attachLifecycleNotifier(
+      ValueNotifier<AppLifecycleState> lifecycleNotifier) {
+    if (_lifecycleNotifier != null) {
+      _lifecycleNotifier!.removeListener(_handleAppLifecycle);
+    }
+    _lifecycleNotifier = lifecycleNotifier;
+    _lifecycleNotifier?.addListener(_handleAppLifecycle);
```
Much better, thank you.

parlough commented on game_template/lib/src/game_internals/level_state.dart:
```diff
@@ -0,0 +1,28 @@
+
+  int get progress => _progress;
+
+  void setProgress(int value) {
+    _progress = value;
+    notifyListeners();
+  }
```
Sounds like the right time to not use a setter then, thanks for the clarification :)

filiph commented:
Thanks for the reviews, @domesticmouse and @parlough! I _think_ I addressed all issues but there are some discussions I left unresolved. Feel free to either close them or continue.

domesticmouse commented:
I'm happy to land this as is, and fast follow changes as required. @parlough is there anything outstanding from your point of view?

parlough commented:
Looks good to me. The required 2.17 changes and anything else can be completed as needed :)

sfshaza2 commented on game_template/android/.gitignore:
```diff
@@ -0,0 +1,13 @@
+GeneratedPluginRegistrant.java
+
+# Remember to never publicly share your keystore.
+# See https://flutter.dev/docs/deployment/android#reference-the-keystore-from-the-app
```
It's SUPER interesting that the anchor in your original URL didn't work. I assume it has to do with using a flutter.dev URL rather than a docs.flutter.dev URL...

```suggestion
# For more info, check out
# [Reference the keystore from the app](https://docs.flutter.dev/deployment/android#reference-the-keystore-from-the-app).
```

filiph replied:
These things (like `.gitignore`) come from Flutter SDK itself. For example: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/app_shared/android.tmpl/.gitignore.

I haven't touched these files. This is what all apps are created with.

sfshaza2 commented on game_template/.metadata:
```diff
@@ -0,0 +1,10 @@
+# This file tracks properties of this Flutter project.
+# Used by Flutter tool to assess capabilities and perform upgrades etc.
```
```suggestion
# The Flutter tool uses this info to assess capabilities and perform upgrades.
```

sfshaza2 reviewed (commented):
I had one or two minor edits. (OK, it's 58.) Otherwise lgtm :D

domesticmouse commented:
@filiph can you please update your PR to merge in the changes we have landed in master?

filiph commented on game_template/README.md:
```diff
@@ -0,0 +1,538 @@
+
+Some general notes: 
+
+- You will want to change the package name of your game
+  before you start with any of the deeper integrations.
+  [Here are the steps](https://stackoverflow.com/a/51550358/1416886)
+  and
+  [here is a tool](https://pub.dev/packages/rename) that automates them.
+- The guides below all assume you already have your game
+  registered in [Google Play Console][] and in Apple's
+  [App Store Connect][].
```
ack

filiph commented on game_template/README.md:
````diff
@@ -0,0 +1,538 @@
+       ? 'ca-app-pub-1234567890123456/1234567890'
+       : 'ca-app-pub-1234567890123456/0987654321';
+   ```
+9. Uncomment the code relating to ads in `lib/main.dart`. You will
+   need to add two imports, too:
````
ack

filiph commented on game_template/README.md:
```diff
@@ -0,0 +1,538 @@
+   `ad_removal`).
+3. Still in Play Console, _activate_ the In-app product.
+   
+To enable in-app purchases on iOS:
```
Fixed

filiph commented:
@domesticmouse - Did this now. The current run fails with "Verify web demos", though. 

I'm going to wait for a bit, and try again later today. But since the web demos are completely separate from `game_template`, and I'm supposed to land the template today, I'm ready to ignore that failure.
